util/audio.py

```diff
@@ -1,31 +1,108 @@
 import os
+import io
 import sox
 import wave
+import opuslib
 import tempfile
 import collections
+import numpy as np
+
 from webrtcvad import Vad
+from util.helpers import LimitingPool
 
 DEFAULT_RATE = 16000
 DEFAULT_CHANNELS = 1
 DEFAULT_WIDTH = 2
 DEFAULT_FORMAT = (DEFAULT_RATE, DEFAULT_CHANNELS, DEFAULT_WIDTH)
 
+AUDIO_TYPE_NP = 'np'
+AUDIO_TYPE_PCM = 'pcm'
+AUDIO_FILE_PREFIX = 'audio/'
+AUDIO_TYPE_WAV = AUDIO_FILE_PREFIX + 'wav'
+AUDIO_TYPE_OPUS = AUDIO_FILE_PREFIX + 'opus'
+LOADABLE_FILE_FORMATS = [AUDIO_TYPE_WAV, AUDIO_TYPE_OPUS]
+
+OPUS_PCM_LEN_SIZE = 4
+OPUS_RATE_SIZE = 4
+OPUS_CHANNELS_SIZE = 1
+OPUS_WIDTH_SIZE = 1
+OPUS_CHUNK_LEN_SIZE = 2
+
+NP_TYPE_LOOKUP = [None, np.int8, np.int16, None, np.int32]
+UNSUPPORTED_TYPE = 'Unsupported audio type: {}'
+
+
+class Sample:
+    def __init__(self, audio_type, raw_data, audio_format=None):
+        self.audio_type = audio_type
+        self.audio_format = audio_format
+        if audio_type in LOADABLE_FILE_FORMATS:
+            self.audio = io.BytesIO(raw_data)
+            self.duration = read_duration(audio_type, self.audio)
+        else:
+            self.audio = raw_data
+            if self.audio_format is None:
+                raise ValueError('For audio type "{}" parameter "audio_format" is mandatory')
+            if audio_type == AUDIO_TYPE_PCM:
+                self.duration = get_pcm_duration(len(self.audio), self.audio_format)
+            elif audio_type == AUDIO_TYPE_NP:
+                self.duration = get_np_duration(len(self.audio), self.audio_format)
+            else:
+                raise ValueError(UNSUPPORTED_TYPE.format(self.audio_type))
+
+    def convert(self, new_audio_type):
+        if self.audio_type == new_audio_type:
+            return
+        if new_audio_type == AUDIO_TYPE_PCM and self.audio_type in LOADABLE_FILE_FORMATS:
+            self.audio_format, audio = read_audio(self.audio_type, self.audio)
+            self.audio.close()
+            self.audio = audio
+        elif new_audio_type == AUDIO_TYPE_NP:
+            self.convert(AUDIO_TYPE_PCM)
+            self.audio = pcm_to_np(self.audio_format, self.audio)
+        elif new_audio_type in LOADABLE_FILE_FORMATS:
+            self.convert(AUDIO_TYPE_PCM)
+            audio_bytes = io.BytesIO()
+            write_audio(new_audio_type, audio_bytes, self.audio_format, self.audio)
+            audio_bytes.seek(0)
+            self.audio = audio_bytes
+        else:
+            raise RuntimeError('Audio conversion from "{}" to "{}" not supported'
+                               .format(self.audio_type, new_audio_type))
+        self.audio_type = new_audio_type
+
+
+def convert_samples(samples, audio_type=AUDIO_TYPE_PCM, processes=None):
+    def convert_sample(sample):
+        sample.convert(audio_type)
+        return sample
+    with LimitingPool(processes=processes) as pool:
+        for current_sample in pool.map(convert_sample, samples):
```

Comment:
This blocks and buffers results until the entire iterable is converted, defeating the purpose of the `yield`. Use:

```python
with LimitingPool(processes=process) as pool:
    yield from pool.imap(convert_sample, samples)
```

Reply:
Oh, I see now that `LimitingPool` calls `Pool.imap`. Please rename it to `LimitingPool.imap` to avoid confusion. (And use `yield from`.)